### PR TITLE
targetRef implementation for RequestAuthentication

### PR DESCRIPTION
**Pull request by keithmattix**
**Please provide a description of this PR:**
For Gateway API Gateways (including Ambient waypoints), `RequestAuthentication` resources should use `targetRef` for binding instead of workloadSelector. One notable change in behavior here is that, if a workloadSelector is specified for a waypoint, _it will not apply_. The same behavior will probably eventually make it's way to all Gateway API gateways, but at present, the functionality is behind a feature flag (defaulted to off) for backwards compatibility

### CODE REVIEW

**Comment by keithmattix** [comment]
Fixes #46846

File: pilot/pkg/features/pilot.go [hunk 1]
```diff
@@ -676,6 +676,9 @@ var (
 	// User should not rely on builtin resource labels, this flag will be removed in future releases(1.20).
 	EnableOTELBuiltinResourceLables = env.Register("ENABLE_OTEL_BUILTIN_RESOURCE_LABELS", false,
 		"If enabled, envoy will send builtin lables(e.g. node_name) via OTel sink.").Get()
+
+	EnableGatewayPolicyAttachmentOnly = env.Register("ENABLE_GATEWAY_POLICY_ATTACHMENT_ONLY", false,
```
Comment (hzxuzhonghu) [inline pilot/pkg/features/pilot.go:680]: we donot need such a flag, as the targetRef is optional.

Comment (keithmattix) [reply pilot/pkg/features/pilot.go:680]: The flag is for whether or not Gateway API gateways will ignore policies NOT attached via a targetRef (according to the Gateway API spec). This semantic change would break existing users so I hid it behind a feature flag

Comment (hzxuzhonghu) [reply pilot/pkg/features/pilot.go:680]: You mean previous users have pa with workload selector, which is applied to waypoint?

If so, since ambient is alpha, we may not need this flag

Comment (keithmattix) [reply pilot/pkg/features/pilot.go:680]: This isn't just for waypoints. TargetRef binding can be used for gateway API gateways too

File: pilot/pkg/model/authentication.go [hunk 2]
```diff
@@ -253,10 +273,39 @@ func getConfigsForWorkload(configsByNamespace map[string][]config.Config,
...
 				}
-				var selector labels.Instance
+				var selector labels.Instance // NOTE: nil/empty selector matches all workloads
 				switch cfg.GroupVersionKind {
 				case gvk.RequestAuthentication:
-					selector = cfg.Spec.(*v1beta1.RequestAuthentication).GetSelector().GetMatchLabels()
+					ra := cfg.Spec.(*v1beta1.RequestAuthentication)
+					targetRef := ra.GetTargetRef()
+					// TODO: Figure out a better way to handle this that doesn't rely on labels
+					gatewayName, isGatewayAPIGateway := workloadLabels[constants.IstioGatewayLabel]
```
[the remark this answers is not in the document]
Comment (keithmattix) [reply pilot/pkg/model/authentication.go:282]: Yeah targetRef is not only for waypoint; also Gateway API ingress gateways

File: pilot/pkg/model/authentication.go [hunk 3]
```diff
@@ -253,10 +273,39 @@ func getConfigsForWorkload(configsByNamespace map[string][]config.Config,
...
+							targetRef.GetName() == gatewayName &&
+							(targetRef.GetNamespace() == "" || targetRef.GetNamespace() == namespace) &&
+							ra.GetTargetRef().GetKind() == gvk.KubernetesGateway.Kind {
+							configs = append(configs, cfg)
+							continue
+						}
+					}
+
+					// Default case
+					selector = ra.GetSelector().GetMatchLabels()
```
[the remark this answers is not in the document]
Comment (keithmattix) [reply pilot/pkg/model/authentication.go:307]: It's a bit more elegant to let it be the default case. Another way to put it is that _because_ it's more often used, we want to handle the exceptions to the rule first

File: pilot/pkg/model/authentication.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/authentication.go]: But i have a doubt fort ingress gateways. why does RA should apply to ingress gateway?

Even if it needs in some case,as we could apply multiple k8s gateway to it, then if a RA match one of them, then should it be applied to this gateway?

[the remark this answers is not in the document]
Comment (keithmattix) [reply pilot/pkg/model/authentication.go]: For Gateway API [policy attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/), the policy (in this case, RA) binds to a specific k8s gateway resource. If there are multiple gateways, the user must create multiple RAs according to the spec. However, our implementation allows for the typical Istio (root) namespace inheritance by default unless the policy attachment only feature flag is turned on

File: pilot/pkg/model/policyattachment.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	GetTargetRef() *v1beta1.PolicyTargetReference
+	GetSelector() *v1beta1.WorkloadSelector
+}
+
+type policyMatch string
+
+const (
+	policyMatchSelector policyMatch = "selector"
+	policyMatchDirect   policyMatch = "direct"
+	policyMatchIgnore   policyMatch = "ignore"
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/policyattachment.go:34]: can you comments on these mode

File: pilot/pkg/model/policyattachment.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func getPolicyMatcher(policyName string, opts workloadSelectionOpts, policy policyTargetGetter) policyMatch {
+	gatewayName, isGatewayAPI := opts.workloadLabels[constants.IstioGatewayLabel]
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/policyattachment.go:38]: I canot see where IstioGatewayLabel is set

File: pilot/pkg/model/policyattachment.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	targetRef := policy.GetTargetRef()
+	if isGatewayAPI && targetRef == nil && policy.GetSelector() != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/policyattachment.go:40]: remove policy.GetSelector() != nil

File: pilot/pkg/model/policyattachment.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+		if opts.isWaypoint || features.EnableGatewayPolicyAttachmentOnly {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/policyattachment.go:41]: The flag is not obvious, it means EnableOnlyPolicyAttachment on gateway api

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:40]: For namespace or root namespace policies, we still want those to apply to waypoints and gateways. We just don't want selector policies to apply

File: pilot/pkg/model/policyattachment.go (hunk 7 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:41]: Do you have ideas for clearer names?

File: pilot/pkg/model/policyattachment.go (hunk 5 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:38]: Lol, there's already another const for it. I'll switch to the preexisting one

File: pilot/pkg/model/policyattachment.go (hunk 4 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go]: Good idea! done

File: pilot/pkg/model/policyattachment.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+type policyMatch string
+
+const (
+	// policyMatchSelector is the default behavior. If the workload matches the policy's selector, the policy is applied
+	policyMatchSelector policyMatch = "selector"
+	// policyMatchDirect is used when the policy has a targetRef, and the workload matches the targetRef.
+	// Note that the actual targetRef matching is done within `getPolicyMatcher`
+	policyMatchDirect policyMatch = "direct"
+	// policyMatchIgnore indicatesd that there is no match between the workload and the policy, and the policy should be ignored
```
Comment (whitneygriffith) [inline pilot/pkg/model/policyattachment.go:37]: typo policyMatchIgnore indicates**d**

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:52]: ? I could not understand such a hack requirement.

Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:52]: Then i would argue why canot we apply policies with selector

File: pilot/pkg/model/policyattachment.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	// policyMatchIgnore indicates that there is no match between the workload and the policy, and the policy should be ignored
+	policyMatchIgnore policyMatch = "ignore"
+)
+
+func getPolicyMatcher(policyName string, opts workloadSelectionOpts, policy policyTargetGetter) policyMatch {
+	gatewayName, isGatewayAPI := opts.workloadLabels[constants.GatewayNameLabel]
+	targetRef := policy.GetTargetRef()
+	if isGatewayAPI && targetRef == nil && policy.GetSelector() != nil {
+		if opts.isWaypoint || features.EnableGatewayPolicyAttachmentOnly {
+			log.Warnf("Ignoring workload-scoped RequestAuthentication %s.%s for gateway %s because it has no targetRef", opts.namespace, policyName, gatewayName)
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/policyattachment.go:54]: nit:  many types including `RequestAuthentication`  here

File: pilot/pkg/model/policyattachment.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:53]: How about EnableSelectorBasedK8sGatewayPolicy,

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM with several comments

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:52]: The key is the different personas. If a platform admin sets a policy on the root mesh namespace, a namespace owner who deploys a waypoint for a particular service account should inherit that policy by default unless they write a conflicting one. This is how Istio works today, right?

File: pilot/pkg/model/policyattachment.go (hunk 7 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:53]: Keep in mind the defaults; because bools have a zero value of false, I picked a name that indicates the opposite of the new behavior. To clarify, if the flag is set to true, then workloadSelector policies that select the gateway label WILL NOT APPLY. Put another way, the flag forces users to apply a targetRef

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:52]: I see no difference with selector

File: pilot/pkg/model/policyattachment.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:53]: The intention is to indicate whether a policy with a selector can apply to a k8s gateway. 
I do not think bool type matters here.

TBH I think `EnableGatewayPolicyAttachmentOnly ` is more clearer than the negative naming `EnableGatewayPolicyAttachmentOnly `.

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:52]: The difference is purely semantic. With a targetRef, it is impossible to write a policy that applies to a request/connection _after_ load balancing (which is undesirable and bloats the config considerably due to Envoy implementation). With workload selector, you can write such policies. While this is ok in sidecars, it's problematic in Ambient which is why targetRef as an API change prevents it. You can find more info in the [Ambient APIs doc](https://docs.google.com/document/d/1Lk1I1qB-XwVWFP0sVcIN5JMJZHjuF3j8lGWtUeDK6pY/edit#heading=h.8nuwse76cdev)

File: pilot/pkg/model/policyattachment.go (hunk 9 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:54]: Ah good catch!

File: pilot/pkg/model/policyattachment.go (hunk 7 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:53]: Did you mean EnableSelectorBasedK8sGatewayPolicy for the first one? The more I think about it, I think I agree with you; will rename

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:53]: Here I see waypoint as a special workload, just like ingress gateway(k8s). Can we disallow bind a policy with workload selector to a ingress gateway? I think no, we cannot.

As a user,  I don't care much about whether the policy is applied before or after loadbalancing. I am not clear what the problem it is for `This makes policies bound to workload selectors problematic.`

File: pkg/config/constants/constants.go [hunk 10]
```diff
@@ -164,7 +164,9 @@ const (
 
 	RemoteGatewayClassName   = "istio-remote"
 	WaypointGatewayClassName = "istio-waypoint"
-	// TODO formalize this API
+
+	// GatewayNameLabel indicates the gateway managing a particular proxy instances. Only populated for Gateway API gateways
+	// TODO: Formalize this API
```
Comment (howardjohn) [inline pkg/config/constants/constants.go:169]: close to merging https://github.com/kubernetes-sigs/gateway-api/pull/1757

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:53]: The only reason workloadSelector worked as an API in the first place is because of the sidecar model: there's always a server sidecar to apply the policy after load balancing. To implement the same thing for Gateway API gateways (including waypoints) requires ugly internal listener logic in Envoy. Instead of carrying that forward, this initiative aligns with Gateway API by explicitly binding policy to the waypoints of the server

Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:53]: I still do not get the persuasive point about this.

@howardjohn @linsun Can you explain the limitation why workload selector is a problem for gateway https://docs.google.com/document/d/1Lk1I1qB-XwVWFP0sVcIN5JMJZHjuF3j8lGWtUeDK6pY/edit

File: pilot/pkg/model/policyattachment.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+type policyTargetGetter interface {
+	GetTargetRef() *v1beta1.PolicyTargetReference
+	GetSelector() *v1beta1.WorkloadSelector
+}
+
+type workloadSelectionOpts struct {
```
Comment (whitneygriffith) [inline pilot/pkg/model/policyattachment.go:31]: Can we export this struct? I need this for a couple of tests outside of the model package. For example: [fuzz_test.go](https://github.com/istio/istio/pull/46560/files#diff-584f7975eb25111ba9a137c0362d4f7cd526bfcc814355af0281989e57724c37)

Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:31]: I don't see why not

File: pkg/config/constants/constants.go (hunk 10 above)
Comment (keithmattix) [reply pkg/config/constants/constants.go:169]: Nice! Will be great to get this standardized

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (howardjohn) [reply pilot/pkg/model/policyattachment.go:53]: The selector vs targetRef is not fundamental or _required_, it is slightly better in a variety of ways that add up to make it worthwhile (IMO).

* Selector is a bit implicit. We have pod, service, gateway, and policy all doing label selectors. They are all tied together, but its all very vague and you can have N:M:X:Y relationship between them which is pretty crazy to understand. TargetRef is much more explicit and cleaner to understand (similar to how we made Gateway:Service binding direct with Gateway API, too).
* TargetRef is the recommended solution for attaching policy to gateways, which is what we are doing. Aside from aligning with other products, there is also plans for tooling, conformance testing, etc around this.

The 'load balancing' discussion is about whether we apply a policy with `app=some-workload` to waypoint. In ambient initial launch, we _did_. In the latter launches, we changed the label selector to be on the gateway (e.g. `istio.io/gateway-name=my-waypoint`). This is a big difference. This decision is in many ways orthogonal from targetRef vs selector, but there is some correlation: the distinction that "workload selector" policies apply to workloads and "target ref" apply to Gateways may help guide users to correctly attaching these policies to the waypoint instead of their workloads. I don't think this is a compelling argument on its own, though.

tl;dr: targetRef is a better API model than workloadSelector and the ecosystem at large is moving towards it. This is simply moving us in that direction. It is not strictly required, more of an ergonomic improvement. The reason its important to do now is that ambient is alpha, so we can make breaking changes more easily.

Comment (hzxuzhonghu) [reply pilot/pkg/model/policyattachment.go:53]: OK, got it. `more of an ergonomic improvement` makes sense to me.

As to `Namespace scope or root scoped policy can apply to gateway/waypoint `, where can we document it.

**Review by hzxuzhonghu — APPROVED** [review]
Great

File: pilot/pkg/model/policyattachment.go (hunk 6 above)
Comment (keithmattix) [reply pilot/pkg/model/policyattachment.go:53]: How about in [this section](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#resource-attachment-and-scaling)? I also need to write up a more detailed doc on gw api

**Review by ericvn — APPROVED** [review]
Thanks for the tests.
